src/main/java/rx/internal/operators/OperatorSampleWithObservable.java

```diff
@@ -57,16 +59,18 @@ public void onNext(U t) {
 
             @Override
             public void onError(Throwable e) {
+                // no need to null check, main is assigned before any of the two gets subscribed
+                main.get().unsubscribe();
                 s.onError(e);
-                unsubscribe();
             }
 
             @Override
             public void onCompleted() {
+                // no need to null check, main is assigned before any of the two gets subscribed
+                main.get().unsubscribe();
```

Comment:
The child subscriber may store some resource coming from `main` and use it `onCompleted`, but `main` will close it in its subscription. E.g.,

```
   @Test
    public void foo() {
        Observable<InputStream> source = Observable.create(new OnSubscribe<InputStream>() {
            @Override
            public void call(Subscriber<? super InputStream> subscriber) {
                try {
                    final InputStream input = new URL("http://www.google.com").openStream();
                    subscriber.add(Subscriptions.create(new Action0() {
                        @Override
                        public void call() {
                            try {
                                input.close();
                            } catch (IOException e) {
                                e.printStackTrace();
                            }
                        }
                    }));
                    Thread.sleep(1000);
                    subscriber.onNext(input);
                    Thread.sleep(1000);
                    subscriber.onCompleted();
                } catch (Exception e) {
                    subscriber.onError(e);
                }
            }
        });
        source.sample(Observable.interval(1500, 100, TimeUnit.MILLISECONDS).take(2)).subscribe(new Subscriber<InputStream>() {

            private InputStream lastInputStream;

            @Override
            public void onCompleted() {
                if (lastInputStream != null) {
                    BufferedReader reader = new BufferedReader(new InputStreamReader(lastInputStream));
                    String line = null;
                    try {
                        while ((line = reader.readLine()) != null) {
                            System.out.println(line);
                        }
                    } catch (IOException e) {
                        e.printStackTrace();
                    }
                }
            }

            @Override
            public void onError(Throwable e) {
                e.printStackTrace();
            }

            @Override
            public void onNext(InputStream inputStream) {
                lastInputStream = inputStream;
            }
        });
    }
```